### PR TITLE
Support custom ssh_command option

### PR DESCRIPTION
**Pull request by weaversam8**
Details are in the original issue #3499.

- Depends on a buildkit update in earthly/buildkit#38
- Adds `ssh_command` option to the Git site config
- Updates Earthfile to support building a buildkit fork with a repo name other than "buildkit"

This PR will need to be updated once earthly/buildkit#38 is merged.

### CODE REVIEW

File: buildcontext/gitlookup.go [hunk 1]
```diff
@@ -615,26 +617,33 @@ func parseGitProtocol(remote string) (string, int) {
...
 	if len(path) > n {
 		subPath = path[n+1:]
 		path = path[:n]
 	}
 	host := path[:strings.IndexByte(path, '/')]
 	gitPath := m.prefix + match[(strings.IndexByte(match, '/')+1):] + m.suffix
 
+	sshCommand = ""
+	if m.sshCommand != "" {
+		sshCommand = m.sshCommand
```
Comment (alexcb) [inline buildcontext/gitlookup.go:641]: very cool to see this was done in a way that allows for custom commands for particular hosts.

**Comment by alexcb** [comment]
Thanks for the PR! I wonder if we could create a new test for this under `tests/git-ssh-server`?

Maybe you could have the server start on a diffferent port, which we already support under https://github.com/earthly/earthly/blob/1b5ec2314eace4c991ef10c31b59a37b9da5b37a/tests/git-ssh-server/setup/Earthfile#L49

you could make a test that's very similar to https://github.com/earthly/earthly/blob/1b5ec2314eace4c991ef10c31b59a37b9da5b37a/tests/git-ssh-server/Earthfile#L338
which will start the server on port 2244

however instead of configuring the port via 

```
earthly --config \$earthly_config config 'git.\"git.example.com\"' '{\"auth\": \"ssh\", \"user\": \"git\", \"port\": 2244, \"strict_host_key_checking\": false}'
```

you could instead config it using the ssh command directly, maybe something like:

```
earthly --config \$earthly_config config 'git.\"git.example.com\"' '{\"auth\": \"ssh\", \"user\": \"git\", \"ssh_command\": "ssh -F /dev/null -p 2244\", \"strict_host_key_checking\": false}'
```

**Comment by alexcb** [comment]
I opened a draft PR under https://github.com/earthly/earthly/pull/3569 to trigger our tests (since they require secrets that are only available to users of the earthly org).

**Comment by alexcb** [comment]
@weaversam8 pointed out (via slack) that there's other locations in the code where the port value is used, so that won't work.

I manually tested it with:

```
git:
  github.com:
    auth: "ssh"
    user: "git"
    ssh_command: "ssh -c badcipher"
```

which correctly shows the `ssh_command` is correctly passed through, since it attempts to use the erroneous cipher:

```
Error: build target: build main: failed to solve: resolve build context for target github.com/earthly/earthly+base: read git-hash: failed to load cache key: error fetching default branch for repository git@github.com:earthly/earthly.git:  exit status 128

git ls-remote --symref git@github.com:earthly/earthly.git HEAD
Unknown cipher type 'badcipher'
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

It seems like we could skip having to stand up an ssh server entirely for this test, and just validate that earthly returns with an error that contains the string `Unknown cipher type 'badcipher'`.

**Comment by weaversam8** [comment]
@alexcb - I just pushed a commit above with a test with your idea (thanks!)

The test is currently failing - it looks like the `GIT CLONE` Earthfile instruction isn't processing the ssh_command properly, so I'll need to fix that. That'll probably be Monday or Tuesday.

**Comment by alexcb** [comment]
I think the test could be simplified -- there shouldn't be a need to startup the sshd server.

instead, you can stick this in `tests/Earthfile`:

```
# test-ssh-command-config tests that we can set a custom ssh_command
# The test intentionally sets a bad cipher type, then tests that buildkit fails, and the failure message
# contains the underlying ssh failure due to attempting to use an invalid cipher
test-ssh-command-config:
    RUN ssh-keygen -b 3072 -t rsa -f /root/dont-steal-my-key -q -N "" -C "rsa-testkey"
    RUN echo "#!/bin/sh
set -ex

# needed to get past the \"no SSH key default forwarded from the client\" error
eval \$(ssh-agent)
ssh-add /root/dont-steal-my-key

earthly --config \$earthly_config config git \"{github.com: {auth: ssh, user: git, strict_host_key_checking: false, ssh_command: ssh -c badcipher}}\"
earthly --config \$earthly_config github.com/earthly/earthly+base 2>&1 | tee output.txt;
" >/tmp/run-earthly-with-bad-ssh-cipher && chmod +x /tmp/run-earthly-with-bad-ssh-cipher
    DO +RUN_EARTHLY --exec_cmd=/tmp/run-earthly-with-bad-ssh-cipher
    RUN cat output.txt | acbgrep "Unknown cipher type 'badcipher'"
```

plus add a `BUILD +test-ssh-command-config` under `ga-no-qemu-group4`.

**Comment by weaversam8** [comment]
@alexcb - That's a smart idea - I left the test in place for now since your test didn't test the `GIT CLONE` option, which was also currently broken. It is true that I don't need to start the server, but I figured it was easier to keep the pattern in the `tests/git-ssh-server` since it already contained a test Earthfile with a `GIT CLONE` instruction.

My latest commit fixes this failing test, so it should be ready to go. Are you okay leaving the more complex test case (since it covers `GIT CLONE`?) Or would you prefer to write a simpler test case that also covers `GIT CLONE` (we'd need to have it load an earthfile somehow.)

**Comment by alexcb** [comment]
I would prefer extending the `badcipher` test, since it'll be faster to execute than the sshd-based test. Our integration tests are quite heavy, so any chance to avoid making them even heavier adds to a better internal developer experience.

I would create two tests in this case:

`test-ssh-command-config-via-remote-target` (which is just renaming my `test-ssh-command-config` test)

and 

`test-ssh-command-config-via-git-clone-earthfile-command`, which is a similar test:

```
test-ssh-command-config-via-git-clone-earthfile-command:
    RUN ssh-keygen -b 3072 -t rsa -f /root/dont-steal-my-key -q -N "" -C "rsa-testkey"

    # or alternatively do a COPY some-test.earthfile Earthfile
    RUN echo "VERSION 0.7
test:
    FROM alpine
    GIT CLONE ...... # the command we are testing
" > Earthfile

    RUN echo "#!/bin/sh
set -ex

# needed to get past the \"no SSH key default forwarded from the client\" error
eval \$(ssh-agent)
ssh-add /root/dont-steal-my-key

earthly --config \$earthly_config config git \"{github.com: {auth: ssh, user: git, strict_host_key_checking: false, ssh_command: ssh -c badcipher}}\"
earthly --config \$earthly_config +test 2>&1 | tee output.txt;
" >/tmp/run-earthly-with-bad-ssh-cipher && chmod +x /tmp/run-earthly-with-bad-ssh-cipher
    DO +RUN_EARTHLY --exec_cmd=/tmp/run-earthly-with-bad-ssh-cipher
    RUN cat output.txt | acbgrep "Unknown cipher type 'badcipher'" # this might have to be changed if the error is output slightly differently.
```

File: tests/Earthfile [hunk 2]
```diff
@@ -1510,6 +1512,47 @@ test-exec-stats:
...
+earthly --config \$earthly_config config git \"{github.com: {auth: ssh, user: git, strict_host_key_checking: false, ssh_command: ssh -c badcipher}}\"
+earthly --config \$earthly_config github.com/earthly/earthly+base 2>&1 | tee output.txt;
+" >/tmp/run-earthly-with-bad-ssh-cipher && chmod +x /tmp/run-earthly-with-bad-ssh-cipher
+    DO +RUN_EARTHLY --exec_cmd=/tmp/run-earthly-with-bad-ssh-cipher
+    RUN cat output.txt | acbgrep "Unknown cipher type 'badcipher'"
+
+test-ssh-command-config-via-git-clone-earthfile-command:
+    RUN ssh-keygen -b 3072 -t rsa -f /root/dont-steal-my-key -q -N "" -C "rsa-testkey"
+
+    # or alternatively do a COPY some-test.earthfile Earthfile
```
Comment (alexcb) [inline tests/Earthfile:1536]: can you remove this comment? this was added to help communicate my idea, not for the final version.

File: tests/Earthfile [hunk 3]
```diff
@@ -1510,6 +1512,47 @@ test-exec-stats:
...
+
+# needed to get past the \"no SSH key default forwarded from the client\" error
+eval \$(ssh-agent)
+ssh-add /root/dont-steal-my-key
+
+earthly --config \$earthly_config config git \"{github.com: {auth: ssh, user: git, strict_host_key_checking: false, ssh_command: ssh -c badcipher}}\"
+earthly --config \$earthly_config +test 2>&1 | tee output.txt;
+" >/tmp/run-earthly-with-bad-ssh-cipher && chmod +x /tmp/run-earthly-with-bad-ssh-cipher
+    DO +RUN_EARTHLY --exec_cmd=/tmp/run-earthly-with-bad-ssh-cipher
+    RUN cat output.txt | acbgrep "Unknown cipher type 'badcipher'" # this might have to be changed if the error is output slightly differently.
```
Comment (alexcb) [inline tests/Earthfile:1554]: has this been tested? if so please remove the comment.

Comment (weaversam8) [reply tests/Earthfile]: yes, I ran the target to test it. my mistake, removed.

File: tests/Earthfile (hunk 2 above)
Comment (weaversam8) [reply tests/Earthfile]: ah, I missed this, thanks. removed.

**Review by alexcb — APPROVED** [review]
looks great, thanks for tackling this!

We had a buildkit update which resulted in a merge conflict, I opened #3598 which resolves the conflicts, I'll tackle anything else that's needed before we can merge it.

**Comment by alexcb** [comment]
This has been merged via #3598; thanks for the contribution.

**Pull request closed without merging**